### PR TITLE
Add odo devfile docs.

### PR DESCRIPTION
**Pull request by yhontyk**
Branches 4.4 and 4.5

### CODE REVIEW

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+----
+  README.md    devfile.yaml    pom.xml        src
+----
+
+. Create a URL to access the deployed component:
++
+----
+ $ odo url create --host apps-crc.testing
+----
++
```
Comment (jboxman) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:55]: Brief introduction to output block here.

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+NOTE: You must use your cluster host domain name when creating the URL.
+
+. Push the component to the cluster:
++
+----
+  $ odo push
+----
++
```
Comment (jboxman) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:69]: Brief introduction to output block.

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  Pushing devfile component myspring
+   ✓  Changes successfully pushed to component
+----
+
+. List the URLs of the component to verify that the component has been pushed successfully:
++
+----
+ $ odo url list
+----
++
```
Comment (jboxman) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:97]: Brief introduction to output block.

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+----
+ Found the following URLs for component myspring
+ NAME              URL                                       PORT     SECURE
+ myspring-8080     http://myspring-8080.apps-crc.testing     8080     false
+----
+
+. View your deployed application using the generated URL:
```
Comment (jboxman) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:104]: s/using/by using/

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+  Executing devfile commands for component myspring
+   ✓  Executing devbuild command "/artifacts/bin/build-container-full.sh" [1m]
+   ✓  Executing devrun command "/artifacts/bin/start-server.sh" [2s]
+
+  Pushing devfile component myspring
+   ✓  Changes successfully pushed to component
+----
+
+. List the URLs of the component to verify that the component has been pushed successfully:
```
Comment (jboxman) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:92]: s/has been/was/

File: modules/developer-cli-odo-listing-available-devfile-components.adoc [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// Module included in the following assemblies:
+//
+// cli_reference/developer_cli_odo/creating-a-java-application-using-devfile
+
+[id="listing-available-devfile-components_{context}"]
+= Listing available devfile components
+
+:FeatureName: Listing available devfile components on a cluster in `{odo-title}`
+include::technology-preview.adoc[leveloffset=+1]
```
Comment (jboxman) [inline modules/developer-cli-odo-listing-available-devfile-components.adoc:9]: We can't include a module within a module.

File: modules/developer-cli-odo-listing-available-devfile-components.adoc [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+With `{odo-title}`, you can display all the components that are available for you on the cluster. Note, that which components are available depends on the configuration of your cluster.
```
Comment (jboxman) [inline modules/developer-cli-odo-listing-available-devfile-components.adoc:11]: s/Note, that which//

>The components that are available depend on the configuration of your cluster.

File: modules/developer-cli-odo-listing-available-devfile-components.adoc [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+= Listing available devfile components
+
+:FeatureName: Listing available devfile components on a cluster in `{odo-title}`
+include::technology-preview.adoc[leveloffset=+1]
+
+With `{odo-title}`, you can display all the components that are available for you on the cluster. Note, that which components are available depends on the configuration of your cluster. 
+
+.Procedure
+
+. To list available devfile components on your cluster, run:
```
Comment (jboxman) [inline modules/developer-cli-odo-listing-available-devfile-components.adoc:15]: For consistency with everything else, how about

> List available devfile components on your cluster:

I've been hamming away at "To [perform task], enter the following command", but not everyone follows that convention.

**Review by jboxman — CHANGES_REQUESTED** [review]
@boczkowska, left comments. Thanks!

**Comment by jboxman** [comment]
@boczkowska, let me know when to look again. Thanks.

File: modules/developer-cli-odo-about-devfiles-in-odo.adoc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+[id="about-devfiles-in-odo"]
+= About devfiles in `{odo-title}`
+
+A devfile is a portable file that describes your development environment. Devfiles are YAML files with a defined https://github.com/redhat-developer/devfile/blob/master/docs/devfile.md[schema].
+With devfiles, you can define a portable developmental environment without the need for reconfiguration.
+
+With devfiles, you can describe your development environment such as the source code, IDE tools, application runtimes, and predefined commands. To learn more about the devfiles, see https://redhat-developer.github.io/devfile/[devfiles documentation].
```
Comment (bobfuru) [inline modules/developer-cli-odo-about-devfiles-in-odo.adoc:7]: s/development environment such as/development environment, such as/

File: modules/developer-cli-odo-about-devfiles-in-odo.adoc [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+
+With `{odo-title}`, you can create components from the devfiles. When creating a component by using the devfiles, `{odo-title}` transforms the devfiles into a workspace consisting of multiple containers that run on OpenShift, Kubernetes, or Docker.
```
Comment (bobfuru) [inline modules/developer-cli-odo-about-devfiles-in-odo.adoc:9]: s/OpenShift/{product-title}

File: modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc [hunk 11]
```diff
@@ -0,0 +1,115 @@
+// Module included in the following assemblies:
+//
+// cli_reference/developer_cli_odo/creating-a-java-application-using-devfile
+
+[id="deploying-a-java-application-using-a-devile_{context}"]
+= Deploying a Java application using a devile
+
+In this section, you will learn how to deploy a sample Java project that uses Maven and Java 8 JDK using devfiles.
```
Comment (bobfuru) [inline modules/developer-cli-odo-deploying-a-java-application-using-a-devile.adoc:8]: s/devfiles/a devfile/

File: cli_reference/developer_cli_odo/creating-applications-by-using-devfiles.adoc [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+include::modules/developer-cli-odo-about-devfiles-in-odo.adoc[leveloffset=+1]
+
+== Creating a Java application by using a devfile
+
+.Prerequisites
+
+* `{odo-title}` is installed. 
+* You must know your ingress domain cluster name. Contact your cluster administrator if you do not know it. For example, `apps-crc.testing` is the cluster domain name for https://access.redhat.com/documentation/en-us/red_hat_codeready_containers/[Red Hat CodeReady Containers]
+* Experimental mode is enabled. To enable experimental mode in odo preferences, run:
```
Comment (bobfuru) [inline cli_reference/developer_cli_odo/creating-applications-by-using-devfiles.adoc:19]: s/experimental mode/Experimental mode/ ? not sure

**Review by bobfuru — COMMENTED** [review]
A few more suggestions, in part related to Jason's review comments. Otherwise, LGTM. :)

File: cli_reference/developer_cli_odo/creating-applications-by-using-devfiles.adoc (hunk 12 above)
Comment (boczkowska) [reply cli_reference/developer_cli_odo/creating-applications-by-using-devfiles.adoc]: I've checked how the team and users refer to it in upstream - indeed, it is capitalized. Fixed it.

**Comment by boczkowska** [comment]
@bobfuru thank you fr the review! I've addressed your suggestions, PTAL when you have time. :relaxed:

**Comment by bobfuru** [comment]
LGTM

**Pull request merged**